stm32/uart: Fix race conditions and clearing status in IRQ handler

dpgeorge opened the pull request:
This is an attempt at an alternative fix to one of the commits in #4599 

The idea is to try and make the IRQ handler more atomic in the operations:
- get current IRQ status flags
- deal with RX character
- clear remaining status flags
- call user handler

On the STM32F4 it's very hard to get this right because the only way to clear IRQ status flags is to read SR then DR, but the read of DR may read some data which should remain in the register until the user wants to read it.  And it won't work to cache the read because RTS/CTS flow control will then not work.  So instead the code will disable interrupts if the DR is full and wait for the user to read it before reenabling the interrupts.

This code is not fully tested and probably doesn't work correctly....

andrewleech commented on ports/stm32/uart.c:
```diff
@@ -689,7 +691,14 @@ int uart_rx_char(pyb_uart_obj_t *self) {
-        return self->uartx->DR & self->char_mask;
+        int data = self->uartx->DR & self->char_mask;
+        if (self->mp_irq_trigger & UART_FLAG_RXNE) {
+            self->uartx->CR1 |= USART_CR1_RXNEIE;
```
FWIW the `UART_RXNE_IT_EN(self->uartx);` macro was used in the other if block just above this to re-enable the interrupt after stalled flow control - it would be worth making these two scenario's consistent in terms of interrupt handling.

andrewleech commented:
I'm inclined to agree that disabling the interrupt while buffer is full is likely the best way to deal with the problem, assuming it can be correctly re-enabled afterwards.

I never had a good test case set up for F4 to check the buffering worked, hence I guess not finding that my last change broke it entirely.   
My test case was basically just the uart/bluetooth connection on my F7 based project - without the linked patch it'd regularly fail / lose data, but is generally stable with the patch.

dpgeorge commented:
> My test case was basically just the uart/bluetooth connection on my F7 based project - without the linked patch it'd regularly fail / lose data, but is generally stable with the patch.

Do you understand exactly why it lost data?  I can't see any race condition in the current master code for the F7.  For the F4 I can understand it could have issues, but F7 seems ok...

andrewleech commented:
> 
> 
> > My test case was basically just the uart/bluetooth connection on my F7 based project - without the linked patch it'd regularly fail / lose data, but is generally stable with the patch.
> 
> Do you understand exactly why it lost data? I can't see any race condition in the current master code for the F7. For the F4 I can understand it could have issues, but F7 seems ok...

I believe it was just the very tiny amount of time between when the data was read from the uart peripheral, to when the flags were read/cleared afterwards. It probably wasn't data loss so much as interrupt flags lost.

dpgeorge commented:
> I believe it was just the very tiny amount of time between when the data was read from the uart peripheral, to when the flags were read/cleared afterwards. It probably wasn't data loss so much as interrupt flags lost.

I still don't understand: the reading of RDR will clear RXNE and ORE.  If IDLE is set after that point but before the code that checks ISR for IDLE it should still pick it up....

andrewleech commented:
Ah, looking at it again, I think it's to do with the initial purpose of that PR - to allow exposing the RXNE interrupt to python.

In the original version, reading of `uartx->DR` is done first, which clears the RXNE flag.
Then when the flags are collected later to expose to python, it's been lost.

dpgeorge commented:
> In the original version, reading of `uartx->DR` is done first, which clears the RXNE flag.
> Then when the flags are collected later to expose to python, it's been lost.

Ah, yes, makes sense!

andrewleech commented:
FWIW I've been running this MR in my builds for over a year now and it has been working well for me, no uart issues seen (on stm32F765, uart used for ble transport)
I've got a recently rebased copy of it at https://github.com/andrewleech/micropython/tree/stm32-uart-irq-fix

dpgeorge committed:
stm32/uart: Fix race conditions and clearing status in IRQ handler.

Prior to this commit IRQs on STM32F4 could be lost because SR is cleared by
reading SR then reading DR.  For example, if both RXNE and IDLE IRQs were
active upon entry to the IRQ handler, then IDLE is lost because the code
that handles RXNE comes first and accidentally clears SR (by reading SR
then DR to get the incoming character).

This commit fixes this problem by making the IRQ handler more atomic in the
following operations:
- get current IRQ status flags
- deal with RX character
- clear remaining status flags
- call user handler

On the STM32F4 it's very hard to get this right because the only way to
clear IRQ status flags is to read SR then DR, but the read of DR may read
some data which should remain in the register until the user wants to read
it.  And it won't work to cache the read because RTS/CTS flow control will
then not work.  So instead the new code disables interrupts if the DR is
full and waits for the user to read it before reenabling the interrupts.

Fixes issue mentioned in #4599 and #6082.

Signed-off-by: Damien George <damien@micropython.org>
